OpenRA.Mods.Common/Traits/Mobile.cs

```diff
@@ -536,31 +225,12 @@ public void SetVisualPosition(Actor self, WPos pos)
 			CenterPosition = pos;
 			self.World.UpdateMaps(self, this);
 
-			// HACK: The submerging conditions must be applied part way through a move, and this is the only method that gets called
-			// at the right times to detect this
-			if (toCell.Layer == CustomMovementLayerType.Subterranean)
-			{
-				var depth = self.World.Map.DistanceAboveTerrain(self.CenterPosition);
-				if (subterraneanToken == ConditionManager.InvalidConditionToken && depth < Info.SubterraneanTransitionDepth && conditionManager != null
-						&& !string.IsNullOrEmpty(Info.SubterraneanCondition))
-					subterraneanToken = conditionManager.GrantCondition(self, Info.SubterraneanCondition);
-			}
-			else if (subterraneanToken != ConditionManager.InvalidConditionToken)
-			{
-				var depth = self.World.Map.DistanceAboveTerrain(self.CenterPosition);
-				if (depth > Info.SubterraneanTransitionDepth)
-				{
-					subterraneanToken = conditionManager.RevokeCondition(self, subterraneanToken);
-
-					// HACK: the submerging animation and sound won't play if a condition isn't defined
-					if (!string.IsNullOrEmpty(Info.SubterraneanTransitionSound))
-						Game.Sound.Play(SoundType.World, Info.SubterraneanTransitionSound);
+			// The first time SetVisualPosition is called is in the constructor before creation, so we need a null check here as well
+			if (notifyVisualPositionChanged == null || !notifyVisualPositionChanged.Any())
```

Comment:
Neglectible: I don't think the `.Any` check is needed. When the collection is empty, we'll just loop zero times.